tower-http/src/classify.rs

```diff
@@ -0,0 +1,170 @@
+//! Tools for classifying responses as either success or failure.
+
+use http::{HeaderMap, Response, StatusCode};
+use http_body::Body;
+use std::{convert::Infallible, marker::PhantomData};
+
+/// Trait for classifying responses as either success or failure.
+pub trait ClassifyResponse {
+    /// The output of doing the classification.
+    ///
+    /// This could be an HTTP status code or some other kind of status.
+    type Output;
+
+    /// The type used to classify the end of streaming responses.
+    type ClassifyEos: ClassifyEos<Output = Self::Output>;
+
+    /// Perform the classification.
+    fn classify_response<B>(
+        &mut self,
+        res: &Response<B>,
+    ) -> ClassifiedNowOrLater<Self::Output, Self::ClassifyEos>
+    where
+        B: Body;
+}
+
+/// Trait for classifying end of streams as either success or failure.
+pub trait ClassifyEos {
```

Comment:
This took me a second to figure out, but maybe define "eos" as prior first use? It doesn't seem to be the _most_ common acronym. Could something like this work?

```suggestion
/// Trait for classifying end of streams ("Eos") as either success or failure.
pub trait ClassifyEos {
```